### PR TITLE
Added max_sample_ arguments

### LINKED ISSUE #10437: [Trainer] add --max_train_samples  --max_val_samples  --max_test_samples
As we were planning to add `--max_train_samples  --max_val_samples  --max_test_samples`  to all examples https://github.com/huggingface/transformers/issues/10423, I thought is there any reason why we don't expand the Trainer to handle that?

It surely would be useful to be able to truncate the dataset at the point of Trainer to enable quick testing.

Another plus is that the metrics can then automatically include the actual number of samples run, rather than how it is done at the moment in examples.

That way this functionality would be built-in and examples will get it for free.

@sgugger

### PR DESCRIPTION
**Pull request by bhadreshpsavani**
# What does this PR do?

Fixes #10437 #10423

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)?  #10437 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Notes:
All the PyTorch-based examples except the below two files will have support for the arguments by adding these changes.
1. The same changes can be implemented for `run_mlm_flax.py` but since I couldn't test the changes I didn't make changes to that file.
2. `run_generation.py`  
* I have reverted the code changes for three TF-based examples since it was giving an error and we want to keep it as it is.
* Test/Predict code addition is still pending. I will do it next.
## review:
@stas00  @sgugger

### CODE REVIEW

File: examples/language-modeling/run_clm.py [hunk 1]
```diff
@@ -340,25 +347,50 @@ def group_texts(examples):
...
+        if data_args.max_train_samples is not None:
+            train_dataset = train_dataset.select(range(data_args.max_train_samples))
+        train_dataset = train_dataset.map(
+            preprocess_function,
+            batched=True,
+            remove_columns=column_names,
+            num_proc=data_args.preprocessing_num_workers,
+            load_from_cache_file=not data_args.overwrite_cache,
+        )
+    if training_args.do_eval:
```
Comment (stas00) [inline examples/language-modeling/run_clm.py:374]:
```suggestion

    if training_args.do_eval:
```

please add a new line between the ifs so that they don't mesh together (same in all other scripts).

File: examples/test_examples.py [hunk 2]
```diff
@@ -64,6 +65,15 @@ def get_setup_file():
     return args.f
 
 
+def get_results(output_dir):
+    results = {}
+    path = os.path.join(output_dir, "all_results.json")
+    if os.path.exists(path):
+        with open(path, "r") as f:
+            results = json.load(f)
+    return results
```
Comment (stas00) [inline examples/test_examples.py:74]: if we are expecting this to always work, then perhaps:
```
else:
    raise ValueError(f"can't find {path}")
```

otherwise `result["eval_accuracy"]` will complain about a missing key, which is not the problem.

**Review by stas00 — APPROVED** [review]
Excellent work, @bhadreshpsavani!

There are a few small tweak requests that I left in the comments.

Thank you!

File: examples/language-modeling/run_clm.py [hunk 3]
```diff
@@ -340,25 +347,50 @@ def group_texts(examples):
         result["labels"] = result["input_ids"].copy()
         return result
 
+    def preprocess_function(examples):
+        examples = tokenizer(examples[text_column_name])
+        return group_texts(examples)
+
```
Comment (sgugger) [inline examples/language-modeling/run_clm.py:353]: This will not work grouped like this as the `group_texts` function relies on the length of tokenized samples. Moreover, in this case, the number of samples is actually the number of elements in the `lm_datasets` (in the version before your PR) as  `group_texts` changes the number of examples.

Therefore, all the preprocessing should be left as is here and the number of samples selected at the end.

File: examples/language-modeling/run_mlm.py [hunk 4]
```diff
@@ -380,12 +394,30 @@ def group_texts(examples):
...
+                raise ValueError("--do_train requires a train dataset")
+            train_dataset = tokenized_datasets["train"]
+            if data_args.max_train_samples is not None:
+                train_dataset = train_dataset.select(range(data_args.max_train_samples))
+            train_dataset = train_dataset.map(
+                group_texts,
+                batched=True,
+                num_proc=data_args.preprocessing_num_workers,
+                load_from_cache_file=not data_args.overwrite_cache,
+            )
```
Comment (sgugger) [inline examples/language-modeling/run_mlm.py:408]: Same comment as before. The selecting should be done after the preprocessing.

File: examples/language-modeling/run_plm.py [hunk 5]
```diff
@@ -358,12 +372,30 @@ def group_texts(examples):
         #
         # To speed up this part, we use multiprocessing. See the documentation of the map method for more information:
         # https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasets.Dataset.map
-        tokenized_datasets = tokenized_datasets.map(
-            group_texts,
-            batched=True,
-            num_proc=data_args.preprocessing_num_workers,
-            load_from_cache_file=not data_args.overwrite_cache,
-        )
+        if training_args.do_train:
```
Comment (sgugger) [inline examples/language-modeling/run_plm.py:375]: Same comment on this script too.

File: examples/question-answering/run_qa.py [hunk 6]
```diff
@@ -360,7 +374,12 @@ def prepare_train_features(examples):
         return tokenized_examples
 
     if training_args.do_train:
-        train_dataset = datasets["train"].map(
+        if "train" not in datasets:
+            raise ValueError("--do_train requires a train dataset")
+        train_dataset = datasets["train"]
+        if data_args.max_train_samples is not None:
+            train_dataset = train_dataset.select(range(data_args.max_train_samples))
+        train_dataset = train_dataset.map(
```
Comment (sgugger) [inline examples/question-answering/run_qa.py:382]: The `prepare_train_features` function will create multiple entries for each example. So we should do a second select after the preprocessing.

File: examples/question-answering/run_qa.py [hunk 7]
```diff
@@ -411,7 +430,12 @@ def prepare_validation_features(examples):
         return tokenized_examples
 
     if training_args.do_eval:
-        validation_dataset = datasets["validation"].map(
+        if "validation" not in datasets:
+            raise ValueError("--do_eval requires a validation dataset")
+        eval_dataset = datasets["validation"]
+        if data_args.max_val_samples is not None:
+            eval_dataset = eval_dataset.select(range(data_args.max_val_samples))
+        eval_dataset = eval_dataset.map(
```
Comment (sgugger) [inline examples/question-answering/run_qa.py:438]: Same for validation.

File: examples/question-answering/run_qa_beam_search.py [hunk 8]
```diff
@@ -373,7 +387,10 @@ def prepare_train_features(examples):
         return tokenized_examples
 
     if training_args.do_train:
-        train_dataset = datasets["train"].map(
+        train_dataset = datasets["train"]
+        if data_args.max_train_samples is not None:
+            train_dataset = train_dataset.select(range(data_args.max_train_samples))
+        train_dataset = train_dataset.map(
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search.py:393]: Same as in the run_qa script, we should do a second select after preprocessing.

File: examples/question-answering/run_qa_beam_search.py [hunk 9]
```diff
@@ -448,7 +465,10 @@ def prepare_validation_features(examples):
         return tokenized_examples
 
     if training_args.do_eval:
-        validation_dataset = datasets["validation"].map(
+        eval_dataset = datasets["validation"]
+        if data_args.max_val_samples is not None:
+            eval_dataset = eval_dataset.select(range(data_args.max_val_samples))
+        eval_dataset = eval_dataset.map(
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search.py:471]: Same for validation.

**Review by sgugger — COMMENTED** [review]
Thanks for diving into this. There is a slight problem with the language modeling examples and the QA examples: for both, the number of samples in the dataset is actually changed in the preprocessing, so we must take more care to have the right number of samples in the final dataset.

In particular, I don't think we can avoid preprocessing the whole dataset in the language modeling examples.

**Comment by stas00** [comment]
Thank you for having a closer look that I did, @sgugger.

Ideally we should have tests that would have caught this

**Comment by bhadreshpsavani** [comment]
Yea @stas00,
We can modify the test cases to consider these three-argument and add assert statements to check the number of samples stored in all_results.json in `test_examples.py`.

We can even remove arguments such as `train_file` and `validation_file` and use data directly with our three sample arguments. If sample test files are not used in any other places we can even remove them from the repository

**Comment by bhadreshpsavani** [comment]
Hi @stas00,

How can we add test cases for this testing? If we check `max_train_samples` and `max_valid_samples` from metrics and add assert statement that might be possible.

**Comment by stas00** [comment]
> How can we add test cases for this testing? If we check `max_train_samples` and `max_valid_samples` from metrics and add assert statement that might be possible.

Yes, that's exactly the idea

**Comment by bhadreshpsavani** [comment]
Hi @stas00,

What should I do if I got this error while using git,
```
$ git push origin argument-addition
To https://github.com/bhadreshpsavani/transformers.git
 ! [rejected]            argument-addition -> argument-addition (non-fast-forward)
error: failed to push some refs to 'https://github.com/bhadreshpsavani/transformers.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

**Comment by bhadreshpsavani** [comment]
I found that I need to use this command `git push -f origin argument-addition` with fast-forward flag.
Thanks, @stas00 I used your rebase script. It's cool! I did it the first time!

**Comment by stas00** [comment]
Some unrelated to your work CI tests were failing so I rebased your PR to master, and then they passed. So you needed to do `git pull` before continuing pushing. and if you already made some changes and `git pull` doesn't work, you normally do:
```
git stash
git pull
git stash pop
```
and deal with any conflicts if any emerge.

In general force-pushing should only be reserved for when a bad mistake was made and you need to revert a bunch of commits.

**Comment by bhadreshpsavani** [comment]
Okay @stas00,
I will be careful while using force push I will use `stash`.
Now I understood

File: examples/language-modeling/run_clm.py [hunk 10]
```diff
@@ -346,19 +361,37 @@ def group_texts(examples):
...
+    if training_args.do_train:
+        if "train" not in tokenized_datasets:
+            raise ValueError("--do_train requires a train dataset")
+        train_dataset = tokenized_datasets["train"].map(
+            group_texts,
+            batched=True,
+            num_proc=data_args.preprocessing_num_workers,
+            load_from_cache_file=not data_args.overwrite_cache,
+        )
```
Comment (sgugger) [inline examples/language-modeling/run_clm.py:373]: I think this map can be dome has before (deleted lines 349 to 354 in the diff) since it's the same for training and validation.

File: examples/language-modeling/run_plm.py [hunk 11]
```diff
@@ -358,12 +372,29 @@ def group_texts(examples):
...
+            if "train" not in tokenized_datasets:
+                raise ValueError("--do_train requires a train dataset")
+            train_dataset = tokenized_datasets["train"].map(
+                group_texts,
+                batched=True,
+                num_proc=data_args.preprocessing_num_workers,
+                load_from_cache_file=not data_args.overwrite_cache,
+            )
```
Comment (sgugger) [inline examples/language-modeling/run_plm.py:383]: Same here.

File: examples/question-answering/run_qa.py [hunk 12]
```diff
@@ -360,13 +374,17 @@ def prepare_train_features(examples):
...
+            raise ValueError("--do_train requires a train dataset")
         train_dataset = datasets["train"].map(
             prepare_train_features,
             batched=True,
             num_proc=data_args.preprocessing_num_workers,
             remove_columns=column_names,
             load_from_cache_file=not data_args.overwrite_cache,
         )
+        if data_args.max_train_samples is not None:
+            train_dataset = train_dataset.select(range(data_args.max_train_samples))
```
Comment (sgugger) [inline examples/question-answering/run_qa.py:387]: We should still do this before the map: the map adds samples but do not reduce their numbers. So in this example, we can speed up preprocessing by doing the `train_dataset = train_dataset.select(range(data_args.max_train_samples))
` before the map so we preprocess at most `max_train_samples` examples, and then once more after the map to make sure we have the right number of examples.

File: examples/question-answering/run_qa.py [hunk 13]
```diff
@@ -411,13 +429,17 @@ def prepare_validation_features(examples):
...
+            raise ValueError("--do_eval requires a validation dataset")
+        eval_dataset = datasets["validation"].map(
             prepare_validation_features,
             batched=True,
             num_proc=data_args.preprocessing_num_workers,
             remove_columns=column_names,
             load_from_cache_file=not data_args.overwrite_cache,
         )
+        if data_args.max_val_samples is not None:
+            eval_dataset = eval_dataset.select(range(data_args.max_val_samples))
```
Comment (sgugger) [inline examples/question-answering/run_qa.py:442]: Same comment here for the validation set.

File: examples/question-answering/run_qa_beam_search.py [hunk 14]
```diff
@@ -373,13 +387,17 @@ def prepare_train_features(examples):
...
+            raise ValueError("--do_train requires a train dataset")
         train_dataset = datasets["train"].map(
             prepare_train_features,
             batched=True,
             num_proc=data_args.preprocessing_num_workers,
             remove_columns=column_names,
             load_from_cache_file=not data_args.overwrite_cache,
         )
+        if data_args.max_train_samples is not None:
+            train_dataset = train_dataset.select(range(data_args.max_train_samples))
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search.py:401]: Same comment as in run_qa

**Review by sgugger — APPROVED** [review]
Thanks for addressing my comments! I have a few more and then it should be ready to be merged :-)

File: examples/language-modeling/run_clm.py (hunk 10 above)
Comment (bhadreshpsavani) [reply examples/language-modeling/run_clm.py:373]: Hi @sgugger,

so we should do it like below
```
lm_datasets = tokenized_datasets.map(
        group_texts,
        batched=True,
        num_proc=data_args.preprocessing_num_workers,
        load_from_cache_file=not data_args.overwrite_cache,
    )
```
and we simply select samples for train and validation?

Comment (sgugger) [reply examples/language-modeling/run_clm.py:373]: Yes. It avoids duplicating the same code this way.

Comment (bhadreshpsavani) [reply examples/language-modeling/run_clm.py:373]: I did this for almost all the examples, I thought preprocessing will be done only if it will be required.
Shall I do these changes for all the examples or mentioned here only?

Comment (sgugger) [reply examples/language-modeling/run_clm.py:373]: For the other examples, you are doing the select before doing the map (to avoid preprocessing all the dataset) so it's not possible to group all the preprocessing together.I think it only applies to the three scripts in run_language_modeling.

**Comment by bhadreshpsavani** [comment]
Hello @stas00 and @sgugger,
I have made the suggested changes
Please let me know if any other changes are required
Thanks

**Comment by sgugger** [comment]
@LysandreJik I think this is ready for final review and merge if you're happy with it.

**Review by LysandreJik — APPROVED** [review]
Great, LGTM!
